metaworld/envs/mujoco/sawyer_xyz/v2/sawyer_door_v2.py

```diff
@@ -93,39 +104,73 @@ def reset_model(self):
 
     def _reset_hand(self):
         super()._reset_hand()
+        self.init_tcp = self.tcp_center
+        self.init_left_pad = self.get_body_com('leftpad')
+        self.init_right_pad = self.get_body_com('rightpad')
+
+    @staticmethod
+    def _reward_grab_effort(actions):
+        return (np.clip(actions[3], -1, 1) + 1.0) / 2.0
+
+    @staticmethod
+    def _reward_pos(obs, theta):
+        hand = obs[:3]
+        door = obs[4:7] + np.array([-0.05, 0, 0])
+
+        threshold = 0.12
+        # floor is a 3D funnel centered on the door handle
+        radius = np.linalg.norm(hand[:2] - door[:2])
+        if radius <= threshold:
+            floor = 0.0
+        else:
+            floor = 0.04 * np.log(radius - threshold) + 0.4
+        # prevent the hand from running into the handle prematurely by keeping
+        # it above the "floor"
+        above_floor = 1.0 if hand[2] >= floor else reward_utils.tolerance(
+            floor - hand[2],
+            bounds=(0.0, 0.01),
+            margin=floor / 2.0,
+            sigmoid='long_tail',
+        )
+        # move the hand to a position between the handle and the main door body
+        in_place = reward_utils.tolerance(
+            np.linalg.norm(hand - door - np.array([0.05, 0.03, -0.01])),
+            bounds=(0, threshold / 2.0),
+            margin=0.5,
+            sigmoid='long_tail',
+        )
+        ready_to_open = reward_utils.hamacher_product(above_floor, in_place)
+
+        # now actually open the door
+        door_angle = -theta
+        a = 0.2  # Relative importance of just *trying* to open the door at all
+        b = 0.8  # Relative importance of fully opening the door
+        opened = a * float(theta < -np.pi/90.) + b * reward_utils.tolerance(
+            np.pi/2. + np.pi/6 - door_angle,
+            bounds=(0, 0.5),
+            margin=np.pi/3.,
+            sigmoid='long_tail',
+        )
 
-        rightFinger, leftFinger = self._get_site_pos('rightEndEffector'), self._get_site_pos('leftEndEffector')
-        self.init_fingerCOM  =  (rightFinger + leftFinger)/2
-        self.reachCompleted = False
+        return ready_to_open, opened
 
     def compute_reward(self, actions, obs):
-        del actions
-        objPos = obs[3:6]
-
-        rightFinger, leftFinger = self._get_site_pos('rightEndEffector'), self._get_site_pos('leftEndEffector')
-        fingerCOM  =  (rightFinger + leftFinger)/2
-
-        pullGoal = self._target_pos
+        theta = self.data.get_joint_qpos('doorjoint')
 
-        pullDist = np.linalg.norm(objPos[:-1] - pullGoal[:-1])
-        reachDist = np.linalg.norm(objPos - fingerCOM)
-        reachRew = -reachDist
+        reward_grab = SawyerDoorEnvV2._reward_grab_effort(actions)
+        reward_steps = SawyerDoorEnvV2._reward_pos(obs, theta)
 
-        self.reachCompleted = reachDist < 0.05
+        reward = sum((
+            2.0 * reward_utils.hamacher_product(reward_steps[0], reward_grab),
```

Comment:
so the one problem that I see happening here is that if you do this then the agent will learn to close the gripper before getting to the object.

Reply:
yep. that's the idea :)